internal/tgz/tgz.go

```diff
@@ -8,60 +8,45 @@ import (
 	"os"
 
 	"github.com/go-git/go-billy/v5"
-	"github.com/go-git/go-billy/v5/util"
+	"github.com/go-git/go-billy/v5/memfs"
 )
 
-const (
-	useDefaultTempDir = ""
-	tmpPrefix         = "tmp-tgz-"
-)
+var MemFactory = func() (billy.Filesystem, error) {
+	return memfs.New(), nil
+}
 
-// Extract decompress a gziped tarball into a new temporal directory
-// created just for this purpose.
+// Extract decompress a gziped tarball into a billy Filesystem.
 //
 // On success, the path of the newly created directory and a nil error
 // is returned.
 //
 // A non-nil error is returned if the method fails to complete. The
-// returned path will be an empty string if no information was extracted
-// before the error and the temporal directory has not been created.
+// returned path will be an empty string if no information was extracted.
 // Otherwise, a non-empty string with the temporal directory holding
 // whatever information was extracted before the error is returned.
-func Extract(fs billy.Filesystem, tgz string) (d billy.Filesystem, err error, cleanup func()) {
-	dirName := ""
-	cleanup = func() {
-		if dirName != "" {
-			_ = os.RemoveAll(dirName)
-		}
-	}
-
-	f, err := fs.Open(tgz)
-	if err != nil {
-		return
-	}
-
+func Extract(tgz billy.File, newFS func() (billy.Filesystem, error)) (fs billy.Filesystem, err error) {
```

Comment:
```suggestion
func Extract(tgz billy.File, fs billy.Filesystem) (path string, err error) {
```

It's probably cleaner to pass the filesystem instead of a func that creates one for multiple reasons. 1) Not all FS types have the same newFS func signature 2) Managing the lifetime of the `newFS` is better done outside of `Extract` and keeps the functionality of `Extract` KISS

This IMO correctly explains the use of `Extract`, a simple example can look like this:
```go
fs := memfs.New()
tgz, _ := os.Open("file.tgz")
path, _ := Extract(tgz, fs)
newFS, _ := fs.Chroot(path)
```

Reply:
Nice shout. :+1: